### PR TITLE
Webpack 2

### PR DESCRIPTION
**Pull request by sokra**

### CODE REVIEW

**Comment by jhnns** [comment]
Woot woot woot!

Hottest changes (taken from the commit log) are:

- `switched to acorn for it's better es6 support`
- `ES6 parser, ES6 modules`
- `disable full dynamic require by default (require(expr))`

and

- `test it too` :wink: 

Did you experience any performance benefits/penalties from switching to acorn? What is actually *breaking* with `2.0.0`?

**Comment by jhnns** [comment]
Btw: I tested this branch in a fairly big project and it worked without changing anything. :+1:

**Comment by sokra** [comment]
> Did you experience any performance benefits/penalties from switching to acorn? What is actually breaking with 2.0.0?

Only the uncool stuff... If you are only using cool stuff, it propably still works...
But there are more changes coming...

**Comment by MoOx** [comment]
Did you take a look to espree that have a full es6 support now ?

**Comment by eugene1g** [comment]
Great. Just tested in our quite complex build - no issues. The only difference was the change in cache hashes, possibly because of the change to the OccurenceOrderPlugin.

File: test/configCases/target/buffer-default/index.js [hunk 1]
```diff
@@ -1,9 +1,10 @@
 require("should");
 
 it("should provide a global Buffer shim", function () {
-	Buffer.should.be.a.Function;
+	Buffer.should.be.an.Function;
```
Comment (akre54) [inline test/configCases/target/buffer-default/index.js]: "a" is the correct usage here since `Function` doesn't start with a vowel.

Comment (jhnns) [inline test/configCases/target/buffer-default/index.js]: ^^

**Comment by gaearon** [comment]
It would be awesome if this was on NPM as `webpack@next`.

**Comment by sokra** [comment]
added `System.import` API for Code Splitting in ES6:

``` js
System.import("./module").then(function(module) { });

// or with context
System.import("./pages/" + pageName).then(function(page) { });
// create one Chunk per page (different to require.ensure)
```

**Comment by chicoxyzzy** [comment]
Cool stuff! BTW what is the difference between first and second example?

**Comment by sokra** [comment]
The second one contains an dynamic expression in the request.

**Comment by ggarek** [comment]
Awesome!

> BTW what is the difference between first and second example?

Curious too

**Comment by ggarek** [comment]
You wrote in code comments: `create one Chunk per page`

Does it mean, that in this case webpack will create chunk for each module that is under `./pages/**/*` glob?

**Comment by sokra** [comment]
> Does it mean, that in this case webpack will create chunk for each module that is under ./pages/**/* glob?

yep.

**Comment by chicoxyzzy** [comment]
But what if one need conditional module loading without splicing chunks but having bundles
```javascript
const module = hasPremiumAccount? 'premium': 'basic';
System.import(`./somePath/${module}`).then(module => {});
```
where the file tree looks like this
```
somePath/premium.js
somePath/basic.js
somePath/core.js
somePath/premiumStuff.js
```

```javascript
// premium.js
import './core';
import './premiumStuff';
// ...
export default somethingCool;
```

```javascript
// basic.js
import './core';
// ...
export default somethingNotSoCool;
```

will Webpack 2 make two or four (or three) chunks in that case?

**Comment by chicoxyzzy** [comment]
I updated my comment above

**Comment by chicoxyzzy** [comment]
Oh now I see Webpack can't know if I want to import `./somePath/premium` or `./somePath/basic`

**Comment by ggarek** [comment]
It is good case and it is pretty specific. IMO, it would be better to have an ability to fine-tune bundler`s behaviour via explicit options, rather then relying on implicit bundler decision based on static analisys or whatever.

**Comment by sokra** [comment]
> it would be better to have an ability to fine-tune bundler`s behaviour via explicit options

`System.import` (like `require.ensure` and AMD `require`) is a Code Split Point. By default this means it's loaded on demand, but the dependency tree can be modified with the optimization plugins. This could mean that a Code Split Point does no longer form a chunk.

These are the option with included plugins:

* `LimitChunkCountPlugin`: Merge chunks when there are too many
* `MinChunkSizePlugin`: Merge chunks when they are too small
* `AggressiveMergingPlugin`: Merge chunks when they shared too much code

A Code Split Point is emitted as normal dependency when both chunks are merged (or child chunk is empty).

**Comment by let4be** [comment]
@sokra today was playing with webpack2 branch and found out that it refuses to execute 2nd entry point if multiple entry points are defined

```
entry: {
	app: [path.resolve('./src/client/main.js')],
	vendor: ['angular', 'angular-ui-router']
},
```

I have all required script references in index.html and load vendor.js first - it perfectly gets executed,
but when I load app.js - code isn't being executed

Any idea?

**Comment by let4be** [comment]
Obvious workaround is not to use CommonsChunkPlugin and multiple entry points for now

**Comment by sokra** [comment]
yep, thanks, it's a bug...

**Comment by SimenB** [comment]
Any timeline on when this is done? Currently converting to ES6, but a split-point is giving us difficulties. from [this](http://webpack.github.io/docs/code-splitting.html#comment-2067018971), I'm guessing I'm waiting for Webpack 2?

**Comment by sokra** [comment]
Sorry don't have a timeline for this. It depends how much of my time I can dedicate to it.

But if you are using webpack 1 and babel-loader you can mix ES6 import with require stuff.

**Comment by SimenB** [comment]
Yeah, for now I just left that single module alone, so no real issue :smile: 
Couldn't get `require.ensure` to work, so I just went back to AMD for that one. It's great that those can be mixed without an issue!

**Comment by saulshanabrook** [comment]
So is it currently not possible to do `System.import('path/' + moduleName)` in Webpack?

It doesn't seem to build right. It builds, but then in the console I get `Uncaught ReferenceError: System is not defined`.

**Comment by sokra** [comment]
> So is it currently not possible to do System.import('path/' + moduleName) in Webpack?

@saulshanabrook It is possible.

Here is an example: https://github.com/webpack/webpack/tree/webpack-2/examples/code-splitting-harmony

**Comment by sokra** [comment]
Just added a pretty new feature to this branch.

For ES6 modules webpack now tracks which exports from modules are imported (used) and which are unused. Unused exports are now not emitted and the minimizer can remove much unused code.

Here is an example: https://github.com/webpack/webpack/tree/webpack-2/examples/harmony-unused

**Comment by Restuta** [comment]
I want to help to test webpack2 using my app, how would I install it with webpack1 so I can switch back and forth fast enough? is there anything like `npm install -g webpack2`?

**Comment by SimenB** [comment]
@sokra Oh, that sounds awesome! Is that applied before or after babel, or other loaders, hits it?

**Comment by nkt** [comment]
@sokra what do you think about introducing new loader option, like `ast`. There is babel, for example. Babel could return AST, instead of code, which allows to webpack do not repeat parsing step for `require/import` lookups and print AST to string by itself.

**Comment by dmr** [comment]
@sokra What can I do to help with merging this?
@gaearon proposed push the new version to npm as ```webpack@next```and I think easier integration to test this out would be great!

**Comment by webpro** [comment]
Would be great to also have Webpack warn about unused imports.

Btw, does anybody know of an easy way/tool to do this?

**Comment by nkt** [comment]
@webpro eslint

**Comment by SimenB** [comment]
Unused `export`s is more interesting, but if webpack can remove them that's a good start :smile: 
Unused imports is just unused variables

**Comment by mattdesl** [comment]
Tree shaking will be an epic feature. 🙌

To take advantage of this, does webpack2 expect your dependencies to use `jsnext:main` for ES6? Or will it expect modules to publish ES6 in the `main` field? Or some combination of that?

**Comment by sokra** [comment]
@mattdesl Currently it will only use `main`, but I'll add `jsnext:main` (or similar) before the release. Btw. you can configurate it with the [`resolve.packageMains` option](http://webpack.github.io/docs/configuration.html#resolve-packagemains).

So you can publish a library to npm which contains ES6 code (referenced by `jsnext:main` (or similar) field) and transpiles or bundled ES5 code (referenced by `main` field).

`jsnext` is propably a bad name, because there could be "next" js after this jsnext. Maybe `es6:main` is better suited, because it has a connection to "ES6 modules".

**Comment by mattdesl** [comment]
Thanks. :smile: 

> jsnext is propably a bad name, because there could be "next" js after this jsnext. 

Agreed. It might be worth writing a spec/standard like the ["browser" field](https://gist.github.com/defunctzombie/4339901) to avoid fragmentation on npm.

**Comment by Restuta** [comment]
@sokra would you mind replying on my comment above, please?

**Comment by sokra** [comment]
> @sokra I want to help to test webpack2 using my app, how would I install it with webpack1 so I can switch back and forth fast enough? is there anything like npm install -g webpack2?

I'll publish a beta version soon...

**Comment by Rich-Harris** [comment]
> `jsnext` is propably a bad name, because there could be "next" js after this jsnext

and

> Agreed. It might be worth writing a spec/standard like the "browser" field to avoid fragmentation on npm.

Came here via [this comment re jsnext:main](https://github.com/mattdesl/eases/pull/2#issuecomment-151155215). I completely agree that the naming of it is a bit unfortunate, though I think it would be really beneficial to have some kind of community-agreed consensus around ES6 module entry points so that ES6-aware tools can do their thing without a configuration burden.

If anyone is interested in helping reach that consensus, there's a [thread on jsnext:main (or alternatives) over here](https://github.com/jsforum/jsforum/issues/5) – I'd love your input. Avoiding fragmentation is the name of the game.

**Comment by sokra** [comment]
So a beta version is publish under the beta tag. Feel free to try it.

Major changes are:

* Many breaking changes in the API of webpack and plugins.
* ES6 modules. Including Code Splitting (System.import) and HMR.
* Chunk loading error handling. (System.import returns a Promise which can fail)

Please give some feedback... :smile:

**Comment by Restuta** [comment]
Is there any gotchas we should know on what are the breaking changes? E.g. I got this error:

```
ypeError: webpack.optimize.OccurenceOrderPlugin is not a function
    at Object.<anonymous> (/Users/Restuta/Projects/rcn.io/code/webpack.config.dev.js:25:5)
    at Module._compile (module.js:425:26)
    at Object.Module._extensions..js (module.js:432:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/Restuta/Projects/rcn.io/code/devServer.js:6:16)
    at Module._compile (module.js:425:26)
    at Object.Module._extensions..js (module.js:432:10)
``` 
and I am not sure what does it mean.

**Comment by chicoxyzzy** [comment]
@Restuta this is expected. This plugin is enabled by default now and you should delete it from plugins array.

**Comment by ArtemGovorov** [comment]
Is there a list of breaking changes in the plugins API somewhere?

**Comment by chicoxyzzy** [comment]
I've got error
`Uncaught ReferenceError: System is not defined`
trying to run example from [here](https://github.com/webpack/webpack/tree/v2.0.0-beta/examples/harmony) on [my boilerplate](https://github.com/chicoxyzzy/react-for-hipsters-boilerplate/tree/webpack-2) (`webpack-2` branch)
steps to reproduce:
- npm i
- npm start
- http://localhost:3000/
open console

**Comment by chicoxyzzy** [comment]
also there is unused file `build.js` in example

**Comment by danieljuhl** [comment]
@Restuta I think OccurenceOrderPlugin has been renamed to OccurrenceOrderPlugin

**Comment by sokra** [comment]
> Is there any gotchas we should know on what are the breaking changes?

I'll publish a more detailed list in the changelog.

**Comment by jhnns** [comment]
> I'll publish a more detailed list in the changelog.

Looking forward to :+1:

**Comment by kentcdodds** [comment]
Yay! Congrats!

**Comment by seetsy** [comment]
Webpack2 official beta!?!?  Thanks @sokra +all contributors!!!

**Comment by ronkorving** [comment]
Very exciting times :) Congrats on the progress.

**Comment by MoOx** [comment]
@sokra I have to admit that a CHANGELOG would be nice, webpack seems to have no changelog or no notes in github release (ps: for github release, you can generate note from a changelog https://github.com/MoOx/github-release-from-changelog ^^)

**Comment by Restuta** [comment]
@sokra I second need for change log with list of depreciations and breaking changes

**Comment by IngwiePhoenix** [comment]
Cooool. Let's all get partying with this beta and make it one amazing WP release! :)

**Comment by kentcdodds** [comment]
While we're suggesting tools for automating the changelog, I recommend [semantic-release](https://github.com/semantic-release/semantic-release) by @boennemann. You integrate it with CI (like travis) and when new code gets into master (WIP branch gets merged), it analyses your commit messages since the last release to determine the kind of release that needs to take place, publishes that release to npm and GitHub complete with a CHANGELOG on GitHub's releases page.

The [commit message convention](https://github.com/ajoslin/conventional-changelog/blob/master/conventions/angular.md) is solid, reasonable, and easy. And there's even a CLI for generating it with a prompt (see [commitizen](http://commitizen.github.io/cz-cli/)).

*Warning... Self promo* ... I have a free egghead.io series that covers how to set all of this up: [How to Write an Open Source JavaScript Library](https://egghead.io/series/how-to-write-an-open-source-javascript-library)

**Comment by jhnns** [comment]
I agree. A changelog with reliable information about breaking changes is vital for a large project like webpack and its ecosystem.

**Comment by webpro** [comment]
Sorry, off-topic as well re: automated releases. To have the git commits be the changelog without changing commit message behavior, [release-it](https://github.com/webpro/release-it) might suit needs without much friction.

**Comment by phaistonian** [comment]
Any chance the beta will make it to NPM?

**Comment by dtothefp** [comment]
@phaistonian agreed would be great to get a `beta` published to NPM. I tried to install this in my project at work from the referencing the github repo but because the package.json version got bumped to `2.0.1-beta` every webpack loader that lists webpack as a peer dependency, which is pretty much all of them I'm assuming, yells and the package won't install.

**Comment by halt-hammerzeit** [comment]
had to hack a lot of google to find any information on "webpack 2 new features" and got here.
tree shaking is cool.

**Comment by klimashkin** [comment]
@sokra I have a quick question about tree-shaking
Assume, that we have folder `/pages` with dozen of modules and `index.js` module in it with re-exporting list of all other modules like
```
export module1 from 'module1';
export module2 from 'module2';
```
And some upper-level script makes `import {module1} from './pages';`. Will `module2` be eliminated from build and from `page/index.js` re-export list with tree-shaking?

**Comment by sokra** [comment]
@klimashkin no. Only the exports of module2 will be omitted.

Removing module2 totally would be an unsafe operation, because module2 could contain side-effects that wouldn't be executed.

I plan to offer is kind of unsafe optimization as opt-in plugin, but there is nothing there yet.

---

In my opinion the ES6 spec failed here as it doesn't allow this optimization... (Would be difficult to apply in a dynamic system, but possible in a static system like webpack)

**Comment by klimashkin** [comment]
You mean only re-export will be omitted, or both - export and re-export but not module itself?

**Comment by sokra** [comment]
> export and re-export but not module itself?

yes

**Comment by sokra** [comment]
Omitting exports of a module allows uglifys dead code elimination to removed side-effect free code.

**Comment by JamesHenry** [comment]
@sokra is there somewhere specific were we should look out for more info on Webpack 2?

I am happy to help start something if not. It would obviously be great for everyone involved if there was somewhere other than this merged PR where knowledge could be shared and you were not responsible for answering ad hoc questions :)

I was already able to start using webpack 2 on my app without any functional problems which is great, but I had to remove it again as the npm errors from things like webpack-dev-server became too much whenever installing new dependencies. I can see @dtothefp had the same experience.

As with the info, would it possible to get webpack2 versions of other "webpack products" like the dev server set up on npm?

PS. Thanks again for all your amazing work on webpack - for me, it is one of the most impressive frontend projects around

**Comment by dtothefp** [comment]
@JamesHenry I agree would be great to have a beginning knowledge share i.e. New stuff and deprecated plugins/syntax. Also as I mentioned earlier a beta version on npm would be useful because for large projects installing dependency directly from github repo is a no go. 

As for your dev server issues switching to webpack middleware and hot middleware may solve your problems, seems to be a more configurable alternative 

https://github.com/webpack/webpack-dev-middleware
https://www.npmjs.com/package/webpack-hot-middleware

**Comment by sokra** [comment]
> It would obviously be great for everyone involved if there was somewhere other than this merged PR where knowledge could be shared and you were not responsible for answering ad hoc questions :)

I'll write a larger article about webpack 2 once it's finished.

PS: Here is more info, for the curious one: https://docs.google.com/document/d/1tRc0MzvRdGK7EbG2LRW8vSyoxKhR_EvRUz3AQRyFZso/edit

**Comment by kentcdodds** [comment]
@sokra, I'm interested in why this was decided:

> webpack differs from the ES6 modules spec: imports are not hoisted

Seems like a bad call to me, but I'm sure I'm out of context.

**Comment by kentcdodds** [comment]
Also, for:

> - add “es6:main” field as default package main
>   - Name?

I may have missed earlier discussion on this, but for the name I thought I'd just bring up the fact that we'll need to revisit this again every year for each new version of the spec. I imagine this has been considered, but I thought I'd just bring it up just in case. May not be a bad thing.

**Comment by phaistonian** [comment]
@sokra Wondering: will the "tree shaking" support result in bundle sizes comparable to the ones Rollup _claims_ to get?

Because, the way I see it, that's their main selling point.

**Comment by sokra** [comment]
> webpack differs from the ES6 modules spec: imports are not hoisted

In this example:

``` js
imports "a";
require("b");
imports "c";
```

The Spec says execution order is: `a`, `c`, `b`.

Currently webpack differs from the spec by using the execution order: `a`, `b`, `c`. I hope this makes incremental transition to ES6 easier.

---

> ill the "tree shaking" support result in bundle sizes comparable to the ones Rollup claims to get?

Sadly not. webpack need to support CJS and AMD too. These styles are pretty dynamic, which requires wrapping in functions to control the execution time. An idea is to add ES6 modules as output target to webpack (still difficult) and run rollbar on the result.

**Comment by phaistonian** [comment]
@sokra Running rollbar on the result sounds the most "logical" way to go, though mashing up 2 bundlers would feel a bit ..dirty.

Is there an estimation of % slower bundling in production will be since tree shaking will be involved? 

Oh, one little thing I noticed (could be because of glitch) is that the `optimize.OccurenceOrderPlugin` fails (not found) in 2.0.

Will re-check on this one later though.

ps: yay for auto applying `NODE_ENV=production` on -p :)

**Comment by danieljuhl** [comment]
@phaistonian Have you seen issue #53 in webpack docs? https://github.com/webpack/docs/issues/53
`OccurenceOrderPlugin` -> `OccurrenceOrderPlugin`

**Comment by dtothefp** [comment]
@phaistonian if you want to read more fine grained details you should have a look here https://github.com/rollup/rollup/issues/219#issuecomment-150842108. @sokra actually pitches in to this conversation on roll up. Seems to me roll up is a good solution if you are only worried about making es6 JS bundles with no extras and will also be a nice tool if you are using http2 and doing only light bundling of commonly grouped modules. For the rest of us who use http1 and also, more importantly, use webpack for many invaluable tools such as scss compilation/local css/public path replacement, image/font handling via file & url loader, patching of non commonjs/AMD non compliant modules, and isomorphic compilation I think webpack will remain an invaluable tool. 

If you are really concerned about boilerplate bloating your code you can look into CommonsChunksPlugin and generate a vendor bundle that you can cache heavily. Not sure how much this helps with the substitution of exports/imports from es6 to commonjs syntax in each file but it does chunk out some webpack runtime as far as I can tell and eliminates potentially bundling twice a library that is imported into multiple bundle entry points

**Comment by sokra** [comment]
> is that the optimize.OccurenceOrderPlugin fails (not found) in 2.0.

No longer needed, is default.

**Comment by phaistonian** [comment]
@dtothefp Thank you for your pointers and input. 

I do use CommonsChunksPlugin and, for what it's worth, I believe Webpack is the most exciting technology we have seen lately (this and React).

It's not about comparing bundlers, it's about being excited as to what could be done (if any) in order to get the best of both worlds in the world you need/love the most (Webpack).

@sokra: Speaking of defaults, perhaps `HotModuleReplacementPlugin` should also come as default (one day) in dev environment.

**Comment by graue** [comment]
Do I need any special configuration to enable `System.import`? Tried it and it just compiles directly to `System['import']`, then fails at runtime with `ReferenceError: System is not defined`.

**Comment by sokra** [comment]
https://gist.github.com/sokra/27b24881210b56bbaff7

**Comment by dtothefp** [comment]
@sokra awesome....thanks for the gist!!!

**Comment by BenoitZugmeyer** [comment]
@sokra it looks awesome!
* I would rewrite the `System.import` example in *Code Splitting with ES6*. Since the `.catch` does not return anything, the `.then` will be called with `undefined` as first argument if the module fails to load.
* s/comsuming/consuming

**Comment by sokra** [comment]
@BenoitZugmeyer thanks, updated it.

**Comment by kentcdodds** [comment]
I've forked it and fixed various spelling and grammatical errors (see changes [here](https://gist.github.com/kentcdodds/d440d0990020c207f751/revisions), in some cases I'm being nit-picky just because I expect this document will be used for documentation in the future). Also changed `CommonJs` to `CommonJS` as that's the most common casing ([from](http://wiki.commonjs.org/wiki/CommonJS) [what](http://requirejs.org/docs/commonjs.html) I [can tell](https://en.wikipedia.org/wiki/CommonJS)).

I had some items of feedback/questions if it's not too late:

- Seems odd that the `--env` flag behaves differently if there's only one vs having multiple. I would prefer that it always results in an object. I feel like keeping it consistent would help reduce confusion.
- I'm unsure of the use/need of `aliasFields` when you can already provide an array of `mainFields`. The document didn't seem to explain it well enough for me at least.
- Rename `moduleExtensions` to `moduleSuffix`. In my mind `extension` maps to a file extension (like `.js` or `.css`). A `suffix` seems to be more appropriate for what this is actually for (this change would also require that `enforceModuleExtension` is changed to `enforceModuleSuffix`).
- `the configuration file resp. the `context` option` I wasn't sure what this phrase is supposed to say. What is `resp.` short for?

Looks great! I'm really excited to try this out :D

**Comment by karlhorky** [comment]
Maybe resp. is short for 'respective of'?

**Comment by sokra** [comment]
> Seems odd that the --env flag behaves differently if there's only one vs having multiple. I would prefer that it always results in an object. I feel like keeping it consistent would help reduce confusion.

I just pass along the value I get from CLI parsing (yargs). Objects are possible because the lib support them (https://www.npmjs.com/package/yargs#dot-notation).
`--env abc` => `"abc"`
`--env` => `true`
`--env 123` => `123`
`--env.hello` => `{hello: true}`
`--env.hello world` => `{hello: "world"}`

> I'm unsure of the use/need of aliasFields when you can already provide an array of mainFields. The document didn't seem to explain it well enough for me at least.

I added an additional sentense.

``` js
"browser": {
  "fs": false,
  "./index.js": "./browser.js"
}
```

**Comment by kentcdodds** [comment]
Thanks for answering my questions @sokra! I think I misunderstood the `--env` flag. I'm good with what you've described :+1:

```diff
// The content of these field is an object where requests to a key are mapped to the corresponding value
// The content of this field is an object where requests to a key are mapped to the corresponding value
```

I hope you're ok with the nit-picking. Let me know if this isn't the time or place to be nit-picky and feel free to ignore my suggestions altogether :-)

Also, what do you think of my other suggestion?

> Rename moduleExtensions to moduleSuffix. In my mind extension maps to a file extension (like .js or .css). A suffix seems to be more appropriate for what this is actually for (this change would also require that enforceModuleExtension is changed to enforceModuleSuffix).

**Pull request merged**